#2426 inconsistent type for user caps

BE-Webdesign opened the pull request:
Fixes #2426, returns an empty object for users who do not have capabilities.

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into develop

BE-Webdesign committed:
#2426-inconsistent-type-for-user-caps

Fixes #2426, returns an empty object for users without capabilities.

BE-Webdesign committed:
Attempt to fix multisite tests.

Attempting to fix multisite tests.

BE-Webdesign committed:
Fix tests for multisite.

Not really sure why this only triggered on multisite.  It seems odd.

danielbachhuber commented on lib/endpoints/class-wp-rest-users-controller.php:
```diff
@@ -517,6 +517,17 @@ public function prepare_item_for_response( $user, $request ) {
 
 		$context = ! empty( $request['context'] ) ? $request['context'] : 'embed';
+
+		if ( 'edit' === $context ) {
```
Let's check like this:

```
if ( isset( $data['capabilities'] ) && empty( $data['capabilities'] ) ) {
    $data['capabilities'] = new stdClass();
}
```

This way, we'll always fix our data regardless of `context`.

BE-Webdesign replied:
Okay, that's how I originally had it. I thought that since it only ever shows up in edit context, it should be in that conditional.  Will revert it.

danielbachhuber commented on lib/endpoints/class-wp-rest-users-controller.php:
```diff
@@ -517,6 +517,17 @@ public function prepare_item_for_response( $user, $request ) {
+
+		if ( 'edit' === $context ) {
+			if ( empty( $data['capabilities'] ) ) {
+				$data['capabilities'] = new stdClass();
```
I'm not sure whether this should actually default to an empty object, or `null`. Can you take a look to see if there's prior art?

BE-Webdesign replied:
I would lean towards empty object as the schema itself is object, or we can change the schema do be array( 'object', 'null' );  I will check to see if there is any info regarding what is better.

danielbachhuber replied:
> Can you take a look to see if there's prior art?

I took a look at the codebase. For the date fields, we use `null` instead of an empty string if there's no date stored. For `parent`, we use `null` if `0` is stored in the database. I think we should probably use `null` instead of an empty object.

danielbachhuber commented on tests/test-rest-users-controller.php:
```diff
@@ -1079,8 +1096,16 @@ protected function check_user_data( $user, $data, $context, $links ) {
+				$this->assertEquals( new stdClass(), $data['capabilities'] );
+			} else {
+				$this->assertEquals( $user->allcaps, $data['capabilities'] );
+			}
```
This doesn't seem quite right — our tests shouldn't be expecting two different types.

BE-Webdesign replied:
It's because associative arrays equal objects in json.  That is why there is the difference.  The weird part is that before I put these changes in, this test only flagged for failing on running multisite.xml.  I agree that this should not be there though, but I am not quite sure how to get around having to include it.

BE-Webdesign committed:
Set capabilities to null when empty

Sets capabilities to null when empty.  This matches with how other empty
values are handled in the API.

BE-Webdesign commented:
Forgot the tests for multisite don't work the same on check_user_data.

BE-Webdesign committed:
Fix test for multisite.

Not sure why this is test only flags on multisite. Makes me somewhat
concerned as I assume it should probably flag on both if it fails.

BE-Webdesign commented:
#review-merge  Not sure why check_user_data only fails on multisite. I think we should probably investigate that some.

rmccue commented on lib/endpoints/class-wp-rest-users-controller.php:
```diff
@@ -517,6 +517,15 @@ public function prepare_item_for_response( $user, $request ) {
 
 		$context = ! empty( $request['context'] ) ? $request['context'] : 'embed';
+
+		if ( isset( $data['capabilities'] ) && empty( $data['capabilities'] ) ) {
```
These checks are redundant; `empty` checks `isset()` internally already.

BE-Webdesign replied:
Not sure what I was thinking exactly, thanks for catching that and pointing it out.

rmccue commented:
(Sorry for the late comment.)

I kinda like an empty object here, since it's an object rather than a scalar value already. It indicates that the user has 0 capabilities rather than not having capabilities at all (if that makes sense?)

Re: multisite failures, that's because on multisite, users don't have any capabilities on a site until they're added to it. On single sites, I believe they always have at least one capability.

BE-Webdesign commented:
> Re: multisite failures, that's because on multisite, users don't have any capabilities on a site until they're added to it. On single sites, I believe they always have at least one capability.

Yeah, I eventually figured that out hehe.

> I kinda like an empty object here, since it's an object rather than a scalar value already.

I would be pro empty object, but I also agree with danielbachuber that we should keep things consistent and most empty values we send through the API come out as null.  If you guys can give me a strong vote either way, I would be glad to revise.

BE-Webdesign committed:
Fixed redundant conditionals, went back to object

Fixed redundant conditionals, went back to returning empty object
instead of null.  This better represents that the user currently has 0
capabilities as opposed to null capabilities.

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into #2426-inconsistent-type-for-user-caps

BE-Webdesign committed:
Fix Tests

Testing fixes.

BE-Webdesign commented:
#reviewmerge - Went back to empty object and fixed conditional checks.

rmccue commented on lib/endpoints/class-wp-rest-users-controller.php:
```diff
@@ -566,6 +566,15 @@ public function prepare_item_for_response( $user, $request ) {
 		$context = ! empty( $request['context'] ) ? $request['context'] : 'embed';
+
+		if ( empty( $data['capabilities'] ) ) {
+			$data['capabilities'] = new stdClass();
```
Rather than using `new stdClass`, can we instead just cast `$user->caps` to an `object` above?

BE-Webdesign replied:
Sure, I like that idea.

rmccue commented on lib/endpoints/class-wp-rest-users-controller.php:
```diff
@@ -817,13 +826,13 @@ public function get_item_schema() {
 				'capabilities'    => array(
 					'description' => __( 'All capabilities assigned to the resource.' ),
-					'type'        => 'object',
+					'type'        => array( 'object', 'null' ),
```
Seems like this patch changes this to actually be just `object`?

BE-Webdesign replied:
I knew I missed something.

rmccue commented on tests/test-rest-users-controller.php:
```diff
@@ -446,6 +446,22 @@ public function test_get_user_invalid_id() {
+
+		$lolz = $this->factory->user->create( array( 'display_name' => 'lolz', 'roles' => '' ) );
+		delete_user_option( $lolz, 'capabilities' );
+		delete_user_option( $lolz, 'user_level' );
```
Nice variable naming. ;)

BE-Webdesign replied:
Do you need me to change it?

rmccue replied:
Given our tests are full of Simpsons references, not an issue. :)

BE-Webdesign replied:
:)

BE-Webdesign committed:
Fixing Schema

Fixing schema of capabilities to match only object.

BE-Webdesign committed:
Merge remote-tracking branch 'refs/remotes/WP-API/develop' into #2426-inconsistent-type-for-user-caps

BE-Webdesign committed:
Clean up - use type casting

Cast caps as an object instead of leveraging array conversion.  Saves
effort on conditionals and makes for cleaner code.

BE-Webdesign commented:
Is this good to go now?
